File: examples/04-decorator/after.ts
/**
 * Example 04 — Decorator, with the pattern.
 *
 * Each concern is its own wrapper, and the composition line says what order they
 * run in. Swap the arguments and the original bug comes back — which is the point:
 * the order is now a visible decision instead of an accident in a loop.
 *
 * Run it:  node examples/04-decorator/after.ts
 */
import assert from 'node:assert/strict';
import { decorate } from '../../packages/ts/src/index.ts';

type Charge = (cents: number) => Promise<string>;

const log: string[] = [];
let attempts = 0;

const rawCharge: Charge = async (cents) => {
  attempts += 1;
  if (attempts < 3) throw new Error('502 from provider');
  return `ch_${cents}`;
};

const withLog = (next: Charge): Charge => async (cents) => {
  log.push(`charging ${cents}`);
  return next(cents);
};

// withRetry knows nothing about payments: it works on anything of this shape.
const withRetry = (times: number) => (next: Charge): Charge => async (cents) => {
  for (let attempt = 0; ; attempt++) {
    try {
      return await next(cents);
    } catch (error) {
      if (attempt >= times) throw error;
    }
  }
};

// Reads in the order it runs: log first, then retries, then the real call.
const charge = decorate(rawCharge, withLog, withRetry(3));

const result = await charge(1999);
console.log(`resultado: ${result}`);
console.log(`intentos:  ${attempts}`);
console.log(`log:       ${log.length} línea → ${log[0]}`);

assert.equal(result, 'ch_1999');
assert.equal(attempts, 3, 'still retried three times');
assert.equal(log.length, 1, 'but logged once');

// Swap the two wrappers and the original bug is back — on purpose. This is the
// part that makes the pattern click.
attempts = 0;
const noisyLog: string[] = [];
const noisyWithLog = (next: Charge): Charge => async (cents) => {
  noisyLog.push(`charging ${cents}`);
  return next(cents);
};
const noisy = decorate(rawCharge, withRetry(3), noisyWithLog);
await noisy(1999);
console.log(`al invertir el orden: ${noisyLog.length} líneas de log`);
assert.equal(noisyLog.length, 3, 'order is a decision, and now a visible one');

// The real payoff: the same layers on another function, with no new code.
attempts = 0;
const rawRefund: Charge = async (cents) => `rf_${cents}`;
const refund = decorate(rawRefund, withLog, withRetry(3));
assert.equal(await refund(500), 'rf_500');
assert.equal(log.length, 2, 'both functions share one retry policy');

console.log('\n✅ un log por llamada, el orden explícito, y las mismas capas reutilizadas');

File: examples/04-decorator/after.py
"""Example 04 — Decorator, with the pattern.

Each concern is its own wrapper, and the composition line says what order they run
in. Swap the arguments and the original bug comes back — which is the point: the
order is now a visible decision instead of an accident in a loop.

Run it:  PYTHONPATH=packages/py python3 examples/04-decorator/after.py
"""

from gof_patterns import decorate

log: list[str] = []
attempts = 0


def raw_charge(cents: int) -> str:
    """Charge the provider."""
    global attempts
    attempts += 1
    if attempts < 3:
        raise RuntimeError("502 from provider")
    return f"ch_{cents}"


def with_log(next_):
    def run(cents: int) -> str:
        log.append(f"charging {cents}")
        return next_(cents)

    return run


def with_retry(times: int):
    """Knows nothing about payments: it works on anything of this shape."""

    def wrapper(next_):
        def run(cents: int) -> str:
            for attempt in range(times + 1):
                try:
                    return next_(cents)
                except RuntimeError:
                    if attempt >= times:
                        raise
            raise AssertionError("unreachable")

        return run

    return wrapper


# Reads in the order it runs: log first, then retries, then the real call.
charge = decorate(raw_charge, with_log, with_retry(3))

result = charge(1999)
print(f"resultado: {result}")
print(f"intentos:  {attempts}")
print(f"log:       {len(log)} línea → {log[0]}")

assert result == "ch_1999"
assert attempts == 3, "still retried three times"
assert len(log) == 1, "but logged once"

# decorate applies functools.wraps, so the identity of the original survives.
print(f"nombre:    {charge.__name__} — {charge.__doc__}")
assert charge.__name__ == "raw_charge"
assert charge.__doc__ == "Charge the provider."

# Swap the two wrappers and the original bug is back — on purpose.
attempts = 0
noisy_log: list[str] = []


def noisy_with_log(next_):
    def run(cents: int) -> str:
        noisy_log.append(f"charging {cents}")
        return next_(cents)

    return run


noisy = decorate(raw_charge, with_retry(3), noisy_with_log)
noisy(1999)
print(f"al invertir el orden: {len(noisy_log)} líneas de log")
assert len(noisy_log) == 3, "order is a decision, and now a visible one"

print("\n✅ un log por llamada, el orden explícito, y wraps preserva la identidad")
